core/src/main/java/io/camunda/migrator/HistoryMigrator.java

```diff
@@ -179,258 +192,330 @@ private void migrateDecisionDefinitions() {
+          .callback(this::migrateHistoricVariable);
     }
+  }
 
-    new Pagination<HistoricVariableInstance>().pageSize(getPageSize())
-        .query(legacyVariableQuery)
-        .callback(legacyVariable -> {
-          String legacyVariableId = legacyVariable.getId();
-          if (!dbClient.checkExists(legacyVariableId)) {
-            LOGGER.debug("Migrating legacy variable with id: [{}]", legacyVariableId);
-            String legacyProcessInstanceId = legacyVariable.getProcessInstanceId();
-            ProcessInstanceEntity processInstance = findProcessInstanceByLegacyId(legacyProcessInstanceId);
-
-            if (processInstance != null) {
-              Long processInstanceKey = processInstance.processInstanceKey();
-              Long scopeKey = findFlowNodeKey(
-                  legacyVariable.getActivityInstanceId()); // TODO does this cover scope correctly?
-              if (scopeKey != null) {
-                VariableDbModel dbModel = variableConverter.apply(legacyVariable, processInstanceKey, scopeKey);
-                variableMapper.insert(dbModel);
-                dbClient.insert(legacyVariableId, legacyVariable.getCreateTime(), dbModel.variableKey(),
-                    HISTORY_VARIABLE);
-                LOGGER.debug("Migration of legacy variable with id [{}] completed.", legacyVariableId);
-              } else {
-                LOGGER.debug("Migration of legacy variable with id [{}] skipped. Activity instance not yet available.",
-                    legacyVariableId);
-              }
-            } else {
-              LOGGER.debug("Migration of legacy variable with id [{}] skipped. Process instance not yet available.",
-                  legacyVariableId);
-            }
+  private void migrateHistoricVariable(HistoricVariableInstance legacyVariable) {
+    String legacyVariableId = legacyVariable.getId();
+    if (shouldMigrate(legacyVariableId)) {
+      LOGGER.debug("Migrating legacy variable with id: [{}]", legacyVariableId);
+      String legacyProcessInstanceId = legacyVariable.getProcessInstanceId();
+
+      // Skip variable if it belongs to a skipped task
+      String taskId = legacyVariable.getTaskId();
+      if (taskId != null && !isMigrated(taskId)) {
+        saveRecord(legacyVariableId, null, IdKeyMapper.TYPE.HISTORY_VARIABLE);
+        LOGGER.debug("Migration of legacy variable with id [{}] skipped. Associated task [{}] was skipped.",
+            legacyVariableId, taskId);
+        return;
+      }
+
+      if (isMigrated(legacyProcessInstanceId)) {
+        ProcessInstanceEntity processInstance = findProcessInstanceByLegacyId(legacyProcessInstanceId);
+        if (isMigrated(legacyVariable.getActivityInstanceId())) {
+          Long processInstanceKey = processInstance.processInstanceKey();
+          Long scopeKey = findFlowNodeKey(
+              legacyVariable.getActivityInstanceId()); // TODO does this cover scope correctly?
+          if (scopeKey != null) {
```

Comment:
I wonder if we have to check the whole hierarchy (PD, PI. activity / task, scope) .
If the PD is not migrated then neither will be thePI or the activity / tasks or scope.
Maybe it is enough to only check the scope (and maybe activity / task